### PR TITLE
Parsing logic for service block

### PR DESCRIPTION
**Pull request by diptanu**

### CODE REVIEW

File: nomad/structs/structs.go [hunk 1]
```diff
@@ -992,6 +992,25 @@ func (tg *TaskGroup) GoString() string {
 	return fmt.Sprintf("*%#v", *tg)
 }
 
+// The ServiceCheck data model represents the consul health check that
+// Nomad registers for a Task
+type ServiceCheck struct {
+	Id       string
+	Name     string
+	Type     string
+	Interval time.Duration
```
Comment (dadgar) [inline nomad/structs/structs.go]: Don't you need a script field and an http endpoint?

File: jobspec/parse.go [hunk 2]
```diff
@@ -401,6 +402,69 @@ func parseTasks(result *[]*structs.Task, list *ast.ObjectList) error {
 			}
 		}
 
+		if o := listVal.Filter("service"); len(o.Items) > 0 {
+			t.Services = make([]structs.Service, len(o.Items))
+			for idx, o := range o.Items {
+				var service structs.Service
```
Comment (mitchellh) [inline jobspec/parse.go]: I would move all of this into a `parseService` function. See how we do it for others.

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -992,6 +992,25 @@ func (tg *TaskGroup) GoString() string {
 	return fmt.Sprintf("*%#v", *tg)
 }
 
+// The ServiceCheck data model represents the consul health check that
```
Comment (dadgar) [inline nomad/structs/structs.go]: Is it worth creating a block of const strings that enumerate the possible check types.
```
const (
    ServiceCheckHTTP = "HTTP"
    ServiceCheckTTL = "TTL"
....
)
```

File: nomad/structs/structs.go [hunk 4]
```diff
@@ -992,6 +992,25 @@ func (tg *TaskGroup) GoString() string {
...
+	Id       string
+	Name     string
+	Type     string
+	Interval time.Duration
+	Timeout  time.Duration
+}
+
+// The Service model represents a Consul service defintion
+type Service struct {
+	Id        string
```
Comment (dadgar) [inline nomad/structs/structs.go]: Put comments on each of the individual fields.

File: nomad/structs/structs.go (hunk 3 above)
Comment (cbednarski) [inline nomad/structs/structs.go]: Can we import these from Consul?

File: api/tasks.go [hunk 5]
```diff
@@ -20,6 +20,25 @@ func NewRestartPolicy() *RestartPolicy {
 	}
 }
 
+// The ServiceCheck data model represents the consul health check that
+// Nomad registers for a Task
+type ServiceCheck struct {
+	Id       string
+	Name     string
+	Type     string
```
Comment (diptanu) [inline api/tasks.go]: Done

File: nomad/structs/structs.go [hunk 6]
```diff
@@ -995,6 +995,53 @@ func (tg *TaskGroup) GoString() string {
...
+	ServiceCheckHTTP   = "http"
+	ServiceCheckTCP    = "tcp"
+	ServiceCheckDocker = "docker"
+	ServiceCheckScript = "script"
+)
+
+// The ServiceCheck data model represents the consul health check that
+// Nomad registers for a Task
+type ServiceCheck struct {
+	Id       string        // If of the check
```
Comment (diptanu) [inline nomad/structs/structs.go]: This will be done on the client side.

File: jobspec/parse.go [hunk 7]
```diff
@@ -452,6 +462,75 @@ func parseTasks(result *[]*structs.Task, list *ast.ObjectList) error {
...
+			return err
+		}
+
+		delete(m, "check")
+
+		if err := mapstructure.WeakDecode(m, &service); err != nil {
+			return err
+		}
+
+		if idx > 0 && service.Name == "" {
```
Comment (dadgar) [inline jobspec/parse.go]: Shouldn't the check be that you have more than one empty name service block. Like I might want one that is has the default name and the rest don't.

File: jobspec/parse.go [hunk 8]
```diff
@@ -452,6 +462,75 @@ func parseTasks(result *[]*structs.Task, list *ast.ObjectList) error {
...
+			return fmt.Errorf("More than one service block is declared, please name each service explicitly")
+		}
+
+		if service.Name == "" {
+			service.Name = fmt.Sprintf("%s-%s-%s", jobName, taskGroupName, task.Name)
+		}
```
Comment (dadgar) [inline jobspec/parse.go]: Can't it be?
``` else {
    service.Name = fmt.Sprintf("%s-%s-%s-%s", jobName, taskGroupName, task.Name, service.Name)
}
```

They should only specify the last part?

File: jobspec/test-fixtures/basic.hcl [hunk 9]
```diff
@@ -45,6 +45,16 @@ job "binstore-storagelocker" {
...
+            service {
+                tags = ["foo", "bar"]
+                port = "http"
+                check {
+                    name = "check-name"
+                    type = "tcp"
+                    interval = "10s"
+                    timeout = "2s"
+                }
+            }
```
Comment (dadgar) [inline jobspec/test-fixtures/basic.hcl]: Probably add some failing case?

File: jobspec/parse.go (hunk 8 above)
Comment (diptanu) [inline jobspec/parse.go]: So we are always forcing the jobname-group-task-name on a service name? I don't mind that, I was wondering user would know what he is doing when he is explicitly name a service.

File: jobspec/parse.go [hunk 10]
```diff
@@ -452,6 +462,81 @@ func parseTasks(result *[]*structs.Task, list *ast.ObjectList) error {
...
+
+		delete(m, "check")
+
+		if err := mapstructure.WeakDecode(m, &service); err != nil {
+			return err
+		}
+
+		if defaultServiceName && service.Name == "" {
+			return fmt.Errorf("More than one service block is declared, please name each service explicitly")
```
Comment (dadgar) [inline jobspec/parse.go]: This has the wrong message.

`return fmt.Errorf("Only one service block may omit name.")

File: jobspec/test-fixtures/basic.hcl (hunk 9 above)
Comment (diptanu) [inline jobspec/test-fixtures/basic.hcl]: Added.

**Comment by dadgar** [comment]
LGTM

File: jobspec/parse_test.go [hunk 11]
```diff
@@ -314,3 +331,20 @@ func TestOverlappingPorts(t *testing.T) {
 		t.Fatalf("Expected collision error; got %v", err)
 	}
 }
+
+func TestIncompleteServiceDefn(t *testing.T) {
+	path, err := filepath.Abs(filepath.Join("./test-fixtures", "incorrect-service-def.hcl"))
+	if err != nil {
+		t.Fatalf("Can't get absoluate path for file: %s", err)
```
Comment (dadgar) [inline jobspec/parse_test.go]: absolute typo

**Pull request merged**